sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala

```diff
@@ -63,25 +63,36 @@ object ExternalCatalogUtils {
     bitSet
   }
 
-  def needsEscaping(c: Char): Boolean = {
+  private final val HEX_CHARS = "0123456789ABCDEF".toCharArray
+
+  @inline final def needsEscaping(c: Char): Boolean = {
     c < charToEscape.size() && charToEscape.get(c)
   }
 
   def escapePathName(path: String): String = {
-    val builder = new StringBuilder()
-    path.foreach { c =>
-      if (needsEscaping(c)) {
-        builder.append('%')
-        builder.append(f"${c.asInstanceOf[Int]}%02X")
-      } else {
-        builder.append(c)
+    var firstIndex = 0
+    val length = path.length
+    while (firstIndex < length && !needsEscaping(path.charAt(firstIndex))) {
+      firstIndex += 1
+    }
+    if (firstIndex == 0) {
+      path
+    } else {
+      val sb = new java.lang.StringBuilder(length + 16)
+      sb.append(path.substring(0, firstIndex))
```

Comment:
Taking another look at the YourKit flamegraph, I think we can actually avoid the allocation overheads of the `.substring` call by instead using the StringBuilder `append` overload that accepts start and end offsets:

```suggestion
      sb.append(path, 0, firstIndex)
```